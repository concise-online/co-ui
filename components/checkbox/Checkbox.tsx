import React, { Component, createRef, DOMAttributes, CSSProperties, ChangeEvent, RefObject } from 'react';
import classNames from 'classnames';
import CheckboxGroup from './CheckboxGroup';
import { runCallback } from '../_utils/function';
import { getPrefixCls } from '../_utils/config';

const prefixCls = getPrefixCls('checkbox');

export interface Props extends Omit<DOMAttributes<HTMLInputElement>, 'onChange'> {
  value?: string;
  className?: string;
  checked?: boolean;
  style?: CSSProperties;
  name?: string;
  disabled?: boolean;
  onChange?: (value: string, e: ChangeEvent<HTMLInputElement>) => void;
  forwardedRef?: RefObject<HTMLInputElement>;
}

class Checkbox extends Component<Props> {
  static Group = CheckboxGroup;
  static isCheckbox = true;
  private ref: RefObject<HTMLInputElement>;
  constructor(props: Props) {
    super(props);

    this.ref = createRef();
    this.handleChange = this.handleChange.bind(this);
  }

  focus() {
    (this.ref.current as HTMLInputElement).focus();
  }

  blur() {
    (this.ref.current as HTMLInputElement).blur();
  }

  handleChange(e: ChangeEvent<HTMLInputElement>) {
    const { disabled, onChange } = this.props;

    if (disabled) return;
    runCallback(onChange, e.target.checked, e);
  }

  render() {
    const { children, className, style, disabled, checked, ...checkboxProps } = this.props;
    const classes = classNames(prefixCls, {
      [`${prefixCls}-checked`]: checked,
      [`${prefixCls}-disabled`]: disabled,
    }, className);

    return (
      <label className={classes} style={style}>
        <span className={`${prefixCls}-element`}>
          <input
            {...checkboxProps}
            type="checkbox"
            checked={checked}
            disabled={disabled}
            onChange={this.handleChange}
            ref={this.ref}
          />
        </span>
        <span className={`${prefixCls}-text`}>
          {children}
        </span>
      </label>
    );
  }
}

export default Checkbox;
