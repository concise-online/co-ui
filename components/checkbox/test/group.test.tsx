import React from 'react';
import { render, mount } from 'enzyme';
import Checkbox from '../index';

describe('checkbox group feature', () => {
  it('renders correctly', () => {
    const wrapper = render((
      <Checkbox.Group value={['foo']}>
        <div>
          <Checkbox value="foo">Foo</Checkbox>
        </div>
        <div>
          <Checkbox value="bar">Bar</Checkbox>
        </div>
      </Checkbox.Group>
    ));

    expect(wrapper).toMatchSnapshot();
  });

  it('change event correctly', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onChange = jest.fn();
    const wrapper = mount((
      <Checkbox.Group value={['foo']} onChange={onChange}>
        <div>
          <Checkbox value="foo">Foo</Checkbox>
        </div>
        <div>
          <Checkbox value="bar">Bar</Checkbox>
        </div>
      </Checkbox.Group>
    ), { attachTo: container });

    wrapper.find('input[value="foo"]').simulate('change');
    expect(onChange).toBeCalled();
  });

  it('set disabled of checkbox group', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onChange = jest.fn();
    const wrapper = mount((
      <Checkbox.Group disabled value={['foo']} onChange={onChange}>
        <div>
          <Checkbox value="foo">Foo</Checkbox>
        </div>
        <div>
          <Checkbox value="bar">Bar</Checkbox>
        </div>
      </Checkbox.Group>
    ), { attachTo: container });

    wrapper.find('input[value="foo"]').simulate('change');
    expect(onChange).not.toBeCalled();
  });
});
