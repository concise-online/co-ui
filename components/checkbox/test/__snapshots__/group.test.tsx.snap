// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox group feature renders correctly 1`] = `
<div
  class="co-checkbox-group"
>
  <div>
    <label
      class="co-checkbox co-checkbox-checked"
    >
      <span
        class="co-checkbox-element"
      >
        <input
          checked=""
          type="checkbox"
          value="foo"
        />
        <span
          class="co-checkbox-inner"
        />
      </span>
      <span
        class="co-checkbox-text"
      >
        Foo
      </span>
    </label>
  </div>
  <div>
    <label
      class="co-checkbox"
    >
      <span
        class="co-checkbox-element"
      >
        <input
          type="checkbox"
          value="bar"
        />
        <span
          class="co-checkbox-inner"
        />
      </span>
      <span
        class="co-checkbox-text"
      >
        Bar
      </span>
    </label>
  </div>
</div>
`;
